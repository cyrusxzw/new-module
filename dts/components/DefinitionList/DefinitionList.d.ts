import React from 'react';
import type { DefinitionListProps } from './DefinitionList.types';
declare const DefinitionList: React.ForwardRefExoticComponent<
  DefinitionListProps & React.RefAttributes<HTMLDListElement>
>;
export { DefinitionList };
