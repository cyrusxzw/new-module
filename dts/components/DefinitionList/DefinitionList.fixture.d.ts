declare const DefinitionListFixture: {
  items: {
    description: string;
    id: string;
    term: string;
  }[];
};
export { DefinitionListFixture };
