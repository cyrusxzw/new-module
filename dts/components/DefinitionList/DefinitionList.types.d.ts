import type { ReactNode } from 'react';
import type { Themes } from '~/types';
declare type DefinitionListItem = {
  description?: ReactNode;
  id: string;
  term?: ReactNode;
};
declare type DefinitionListProps = {
  className?: string;
  hasBottomBorder?: boolean;
  isVisible?: boolean;
  items?: DefinitionListItem[];
  theme?: Themes;
};
export type { DefinitionListProps };
