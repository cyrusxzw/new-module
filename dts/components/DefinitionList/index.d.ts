import { DefinitionList } from './DefinitionList';
export { DefinitionList };
