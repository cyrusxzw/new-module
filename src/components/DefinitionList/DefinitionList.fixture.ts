const DefinitionListFixture = {
  items: [
    {
      description: 'Proin vulputate eros quis lectus pretium gravida.',
      id: '1',
      term: 'Aliquam',
    },
    {
      description: 'Pulvinar nulla enim, eu elementum lacus vehicula sed.',
      id: '2',
      term: 'Etiam eu augue quam',
    },
    {
      description: 'Sed ac ipsum vel risus aliquam rutrum.',
      id: '3',
      term: 'Duis at elit',
    },
  ],
};

export { DefinitionListFixture };
