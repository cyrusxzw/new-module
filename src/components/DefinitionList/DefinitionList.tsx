import React, { Fragment, forwardRef } from 'react';
import cx from 'classnames';
import { isObjectPopulatedArray } from '~/utils/objects';
import styles from './DefinitionList.module.css';
import type { DefinitionListProps } from './DefinitionList.types';

const DefinitionList = forwardRef<HTMLDListElement, DefinitionListProps>(
  function DefinitionListRef(
    {
      className,
      hasBottomBorder = false,
      isVisible = true,
      items = [],
      theme = 'dark',
    },
    ref,
  ) {
    if (!isObjectPopulatedArray(items)) {
      return null;
    }

    const classSet = cx(
      styles.base,
      styles[theme],
      { [styles.hasBottomBorder]: hasBottomBorder },
      className,
    );
    const termClassSet = cx(styles.term, { [styles.slideIn]: isVisible });
    const descriptionClassSet = cx(styles.description, {
      [styles.slideIn]: isVisible,
    });

    return (
      <dl
        className={classSet}
        data-testid="data-testid-DefinitionList"
        ref={ref}
      >
        {items
          .filter(({ description, term }) => description || term)
          .map(({ description, id, term }) => (
            <Fragment key={id}>
              <dt className={termClassSet}>{term}</dt>
              <dd className={descriptionClassSet}>{description}</dd>
            </Fragment>
          ))}
      </dl>
    );
  },
);

export { DefinitionList };
