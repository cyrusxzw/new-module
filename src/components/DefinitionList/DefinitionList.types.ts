import type { ReactNode } from 'react';
import type { Themes } from '~/types';

type DefinitionListItem = {
  description?: ReactNode;
  id: string;
  term?: ReactNode;
};

type DefinitionListProps = {
  className?: string;
  hasBottomBorder?: boolean;
  isVisible?: boolean;
  items?: DefinitionListItem[];
  theme?: Themes;
};

export type { DefinitionListProps };
