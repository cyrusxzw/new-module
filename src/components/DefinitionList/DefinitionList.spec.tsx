import React from 'react';
import { render, screen } from '@testing-library/react';
import { DefinitionList } from './DefinitionList';
import { DefinitionListFixture } from './DefinitionList.fixture';

describe('<DefinitionList />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <DefinitionList items={DefinitionListFixture.items} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should return `null` if the `items` prop is not a populated array', () => {
    render(<DefinitionList items={[]} />);

    const child = screen.queryByTestId(/data-testid-DefinitionList/i);

    expect(child).not.toBeInTheDocument();
  });
});
