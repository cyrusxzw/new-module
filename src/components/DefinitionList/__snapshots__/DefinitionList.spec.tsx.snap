// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DefinitionList /> should render base component correctly 1`] = `
<div>
  <dl
    class="base dark"
    data-testid="data-testid-DefinitionList"
  >
    <dt
      class="term slideIn"
    >
      Aliquam
    </dt>
    <dd
      class="description slideIn"
    >
      Proin vulputate eros quis lectus pretium gravida.
    </dd>
    <dt
      class="term slideIn"
    >
      Etiam eu augue quam
    </dt>
    <dd
      class="description slideIn"
    >
      Pulvinar nulla enim, eu elementum lacus vehicula sed.
    </dd>
    <dt
      class="term slideIn"
    >
      Duis at elit
    </dt>
    <dd
      class="description slideIn"
    >
      Sed ac ipsum vel risus aliquam rutrum.
    </dd>
  </dl>
</div>
`;
